tools/d2g/d2gtest/schemas/test_suites.yml

```diff
@@ -74,3 +58,59 @@ properties:
               # schemas and we prefer not to keep both json and yml versions in
               # the repo
               type: object
+      taskDefTests:
+        title: Task definition test cases
+        description: |-
+          The test cases which this test suite contains
+        type: array
+        items:
+          title: Task definition test case
+          description: |-
+            A test case contains a static input Docker Worker task definition, and an
+            expected Generic Worker task definition output. The Docker Worker task definition
+            is converted by d2g to a Generic Worker task definition. The test is successful
+            if the generated Generic Worker task definition exactly matches the Generic
+            Worker task definition in the test case.
+          type: object
+          required:
+            - name
+            - description
+            - dockerWorkerTaskDefinition
+            - genericWorkerTaskDefinition
+          additionalProperties: false
+          properties:
+            name: {$ref: '#/definitions/caseName'}
+            description: {$ref: '#/definitions/caseDescription'}
+            dockerWorkerTaskDefinition:
+              # validation of this object is handled in test harness, rather
+              # than using $ref, since gojsonschema can't interpret yaml
+              # schemas and we prefer not to keep both json and yml versions in
+              # the repo
+              type: object
+            genericWorkerTaskDefinition:
+              # validation of this object is handled in test harness, rather
+              # than using $ref, since gojsonschema can't interpret yaml
+              # schemas and we prefer not to keep both json and yml versions in
+              # the repo
+              type: object
```

Comment:
It will be easier if you embed all the properties that the d2g test cases include here (such as taskQueueId, payload, scopes, ....). Then you won't need all the unmarshaling in the code. Note, this only pertains to the test case data, so you don't need to worry about task definition properties changing over time, since the input/output pairs are frozen in the tests (so the schema only needs to include the properties that the tests actually contain). Make sure to include "additionalProperties: false" though, so that the generated go code isn't full of json.RawMessage. I would even add an optional (non-required) property "foo", e.g. type object, and create a single test case that includes it with some value, to make sure that it is passed through verbatim to the translated task definition. Only one test needs to test this functionality. Obviously, document "foo" as a fake property for the purposes of testing that unknown properties get passed through.

Incidentally, it occurs to me that when converting tasks, a task definition may include provisionerId/workertype or taskQueueId, but doesn't need to include both. If we translate tasks, I guess we should check if both exist, that they are consistent with each other, and if only one exists, we use that one, and not the other when passing to the Scopes conversion routine. Because if we assume it is present, and then pass through an empty string, we'll get the wrong scope. Just occurred to me. That is also something we can cover with a test case. Normally generic-worker doesn't care, because it gets task definition from Queue service, and that will include both, but if someone is converting tasks definitions that they create themself (e.g. from taskgraph) they might have one of the properties, but not the other, which is valid for taskcluster, but d2g wouldn't understand in its current incarnation.

Reply:
I've addressed your second paragraph's concern in https://github.com/taskcluster/taskcluster/pull/6618/commits/a18262d3e739c8230558b0fdf0211a96791390ec.

Regarding your first paragraph - there are tradeoffs to doing this. The only field accessed in the d2g tests is the `payload` and that's to validate against the worker's payload schema. So that's two unmarshals.

If the `payload` field is added here, I'll then need to change:
  * marshal the field into JSON to do the validation steps I mentioned above
  * https://github.com/taskcluster/taskcluster/blob/a18262d3e739c8230558b0fdf0211a96791390ec/tools/d2g/d2gtest/d2g_test.go#L151 I'll no longer be able to pass the docker worker task definition in as-is. I'll need to marshal the docker worker task definition into JSON before passing to `d2g.ConvertTaskDefinition()`

Keeping it as-is keeps the test case method for task definitions super straight forward: https://github.com/taskcluster/taskcluster/blob/a18262d3e739c8230558b0fdf0211a96791390ec/tools/d2g/d2gtest/d2g_test.go#L147-L170

Let me know what you think! Thanks Pete

Reply:
And in https://github.com/taskcluster/taskcluster/pull/6618/commits/77829a2beaa2b93bc7d5379f6cc25783589fdf00, I just added another test to ensure new fields are translated verbatim through to the resulting generic worker task definition.